compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -731,11 +731,21 @@ impl<'tcx> Cx<'tcx> {
                 });
                 ExprKind::Loop { body }
             }
-            hir::ExprKind::Field(source, ..) => ExprKind::Field {
-                lhs: self.mirror_expr(source),
-                variant_index: FIRST_VARIANT,
-                name: self.typeck_results.field_index(expr.hir_id),
-            },
+            hir::ExprKind::Field(source, ..) => {
+                let mut kind = ExprKind::Field {
+                    lhs: self.mirror_expr(source),
+                    variant_index: FIRST_VARIANT,
+                    name: self.typeck_results.field_index(expr.hir_id),
+                };
+                let nested_field_tys_and_indices =
+                    self.typeck_results.nested_field_tys_and_indices(expr.hir_id);
+                for &(ty, idx) in nested_field_tys_and_indices {
+                    let expr = Expr { temp_lifetime, ty, span: source.span, kind };
+                    let lhs = self.thir.exprs.push(expr);
+                    kind = ExprKind::Field { lhs, variant_index: FIRST_VARIANT, name: idx };
+                }
```

Comment:
So I wonder if making `field_index` be something like a `[2, 0, 0]` here wouldn't be better because it forces any users of `field_index` to deal with the complexity that if you want to find the actual type being referred to, then you need to consider the nested indices.

There are a handful of places that use `field_index` and it's possible those would need updated to work correctly with anonymous ADTs anyway. This would simplify where you're writing field indices above too.

Reply:
> So I wonder if making `field_index` be something like a `[2, 0, 0]` here wouldn't be better because it forces any users of `field_index` to deal with the complexity that if you want to find the actual type being referred to, then you need to consider the nested indices.

Yes, indeed, it was my first attempt. Then I found that the intermediate anonymous ADTs should also be cached in `typeck_results` so that we can reuse them during field accessing, pattern matching, or other cases.

Reply:
I think we probably should do this differently - both because I feel like there's a risk of `field_indices` being used in the compiler but not `nested_field_indices` and that being a source of bugs, but also that the way we create the nested expressions here avoids some of the additional logic that `mirror_expr` has (like wrapping things in `ExprKind::Scope`). I wonder if we avoid this entirely by lowering to nested `ExprKind::Field` in the HIR rather than in the THIR?

Reply:
> both because I feel like there's a risk of `field_indices` being used in the compiler but not `nested_field_indices` and that being a source of bugs, but also that the way we create the nested expressions here avoids some of the additional logic that `mirror_expr` has (like wrapping things in `ExprKind::Scope`).

Agree. 

> I wonder if we avoid this entirely by lowering to nested `ExprKind::Field` in the HIR rather than in the THIR?

Do you mean lowering the nested `ExprKind::Field` here?

https://github.com/rust-lang/rust/blob/6ad80111673c338044093832c950932c2621c598/compiler/rustc_ast_lowering/src/expr.rs#L247-L249

But we don't know the exact type of the `el` here, I'm afraid we should do it after typeck.

---

I wonder if we could expand those nested HIR expressions in place before lowering them to THIR? But I'm not sure if the HIR is still mutable during that pass.

Reply:
Ah, yeah, I guess that isn't particularly feasible without knowing more about the types.

Reply:
Another advantage of lowering them to THIR is that THIR uses `FieldIdx` instead of the concrete field name (or symbol), which avoids lots of distinguishing works in dealing with the almost-identical unnamed fields `_` (they should probably be renamed to `f._1` in HIR regarding their indices in the ADTs).

Reply:
Let me try implementing pattern-matching and other functions that use `field_index`, and we can discuss again the best way to lower the nested fields.

Reply:
> Let me try implementing pattern-matching and other functions that use `field_index` in a few days, and then we can discuss again the best way to lower the nested fields.

Ping me when you have - I don't want to keep this in review limbo for too long, it's largely good and we can always tweak later, better to merge and make progress.